Add mechanism to react to termination signals

AthenaEryma opened the pull request:
This commit adds a mechanism to be exposed internally to allow plugins/modules to react to termination signals via SPI.

I'm not entirely sure how to test this - normally I'd use an `ESIntegTestCase` and supply a stubbed version of the plugin but that's more difficult with SPI.

Supercedes https://github.com/elastic/elasticsearch/pull/95518

AthenaEryma committed:
Add mechanism to react to termination signals

This commit adds a mechanism to be exposed internally
to allow plugins/modules to react to termination signals
via SPI.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

AthenaEryma committed:
I thought I committed the spotless changes already

pgomulka commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -1591,6 +1600,24 @@ private Node stop() {
     public synchronized void close() throws IOException {
+        CountDownLatch terminationHandlersComplete = new CountDownLatch(terminationHandlers.size());
+        terminationHandlers.forEach(handler -> {
+            // This AtomicBoolean is used to ensure one plugin calling its runnable over and over won't run down the CountDownLatch early
```
out of curiosity, when can this happen that one plugin is running continuously (over and over) the handleTermination and it's runnable?

gwbrown replied:
I'll be removing this mechanism in favor of verifying that we have only one of these handlers anyway. This is more of a guard against anything doing it accidentally rather than something I expect to actually happen - if I were keeping this mechanism here I'd probably put an `assert` here to make it clear that multiple calls are not expected.

rjernst reviewed (commented):
While I understand the usual caution with blocking threads, I do not think the async handling here is necessary.

The JVM shutdown hook has its own dedicated thread. So we are fine to block in Node.close().

Additionally, we should only have one of these termination handlers, which calls and blocks on the shutdown API. So we should be fine to have the termination handler be a blocking method, and have a single callback instead of many (and we can assert only one plugin implements TerminationHandlerProvider)

stu-elastic commented on server/src/main/java/org/elasticsearch/node/internal/TerminationHandler.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @param done This runnable should be invoked when this handler has finished preparing the node for shutdown.
+     */
+    void handleTermination(Runnable done);
```
When would a plugin implement this instead of `ShutdownAwarePlugin`'s `signalShutdown`?

AthenaEryma committed:
Merge branch 'main' into shutdown/termination-spi-real

AthenaEryma committed:
Address review feedback

Specifically, enforce one implementing plugin and use a blocking API. Also, enhance Javadoc.

stu-elastic commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -1591,6 +1610,8 @@ private Node stop() {
     // In this case the process will be terminated even if the first call to close() has not finished yet.
     @Override
     public synchronized void close() throws IOException {
+        Optional.ofNullable(terminationHandler.get()).ifPresent(TerminationHandler::handleTermination);
```
Why call this at close?  The javadoc in TerminationHandler indicates `handleTermination` is a result of a signal from the OS.  That handling occurs in [Elasticsearch.java](https://github.com/elastic/elasticsearch/blob/cbcbd01d5e78026d12b9ea29834f39e45def9a5a/server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java#L442).

gwbrown replied:
Plumbing, basically - we don't have a `PluginsService` in `Elasticsearch.java` to get the termination handlers. We could potentially lift the construction of `PluginsService` out of `Node` into `Elasticsearch`, or expose the termination handler as a separate method on `Node` and call it separately if you think that would be better.

pgomulka commented on server/src/main/java/org/elasticsearch/node/internal/TerminationHandlerProvider.java:
```diff
@@ -0,0 +1,33 @@
+ *
+ * <p>Note that this class is mostly for plumbing - translating a low-level signal received by a node process into a higher-level set
+ * of operations. Logic to respond to planned changes in cluster membership should use Node Shutdown primitives instead, see
+ * {@link org.elasticsearch.plugins.ShutdownAwarePlugin} for lower-level plugin operations and
```
not related to this PR, but have we considered moving `ShutdownAwarePlugin` into a stable plugin lib?
It would require the additional translation for SingleNodeShutdownMetadata (it depends on Writeable) though

gwbrown replied:
I'm not aware of any discussions around moving `ShutdownAwarePlugin` into a stable plugin lib. The Node Shutdown API is a bit weird in general since it's technically only supported on Cloud.

rjernst commented on server/src/main/java/org/elasticsearch/node/internal/TerminationHandlerProvider.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @return A list of {@link  TerminationHandler}s
+     */
+    Collection<TerminationHandler> handlers();
```
Why do we need multiple handlers? Shouldn't serverless have exactly one, which calls and then blocks on the shutdown API?

gwbrown replied:
Whoops, I missed this one - I was too focused on the option of there being multiple plugins, I forgot about this. I've made this return a single handler.

rjernst reviewed (commented):
Thanks for the updates, this is looking good. I have one last question.

AthenaEryma committed:
Only a single termination handler, I mean it this time

AthenaEryma committed:
Move handler to a method on Node, per Stu's feedback

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/termination-spi-real

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/termination-spi-real

rjernst commented on server/src/main/java/org/elasticsearch/node/internal/TerminationHandlerProvider.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return A list of {@link  TerminationHandler}s
+     */
+    TerminationHandler handlers();
```
nit: handler, singular

gwbrown replied:
Fixed. Also fixed the Javadoc on this method. So many places to update.

Thanks for all the reviews and feedback, it was greatly useful!

rjernst reviewed (approved):
LGTM. I suspect we may still need this to occur in close, as that is what ESIntegTestCase would call, but it would only be a problem in serverless, so we can change as a followup if necessary.

AthenaEryma committed:
Again, just one handler

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/termination-spi-real

AthenaEryma committed:
Again, just one handler (again)

stu-elastic commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -1670,6 +1690,18 @@ public synchronized void close() throws IOException {
+     * Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown
+     * logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}.
+     */
+    public void prepareForClose() {
```
Nice api and docs.

stu-elastic reviewed (approved):
Nicely done.

droberts195 commented on server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java:
```diff
@@ -444,6 +444,7 @@ private static void shutdown() {
             return; // never got far enough
         }
         var es = INSTANCE;
+        es.node.prepareForClose();
```
3 lines further down there's a check for whether `es.node` is `null` before using it. Either that's not necessary - quite possible as I don't know the code well - or else this line needs to be guarded too.

Also, should this method be inside the `try` block? If it throws an exception then the `finally` block won't run, which presumably means Elasticsearch won't exit.

rjernst replied:
The null check is outdated; it's not possible (I think it was at an earlier time). The INSTANCE null check is what guards early termination. However, moving this into the try though is important, so that any error while shutting down is properly logged and the latch counted down to allow the process to exit.
